network/proc/findpid.go

```diff
@@ -5,99 +5,100 @@ package proc
 func GetPID(socketInfo socket.Info) (pid int) {
-	pidsByUserLock.Lock()
-	defer pidsByUserLock.Unlock()
-
-	if socketInfo.GetPID() != socket.UnidentifiedProcessID {
-		return socketInfo.GetPID()
+	// Get currently assigned PID to the socket info.
+	socketInfoLock.RLock()
+	currentPid := socketInfo.GetPID()
+	socketInfoLock.RUnlock()
+
+	// If the current PID already is valid (ie. not unidentified), return it immediately.
+	if currentPid != socket.UnidentifiedProcessID {
+		return currentPid
 	}
 
+	// Find PID for the given UID and inode.
 	pid = findPID(socketInfo.GetUID(), socketInfo.GetInode())
+
+	// Set the newly found PID on the socket info.
+	socketInfoLock.Lock()
 	socketInfo.SetPID(pid)
+	socketInfoLock.Unlock()
+
+	// Return found PID.
 	return pid
 }
 
 // findPID returns the pid of the given uid and socket inode.
-func findPID(uid, inode int) (pid int) { //nolint:gocognit // TODO
+func findPID(uid, inode int) (pid int) {
+	socketName := fmt.Sprintf("socket:[%d]", inode)
 
-	pidsUpdated := false
+	for i := 0; i <= lookupRetries; i++ {
+		var pidsUpdated bool
 
-	// get pids of user, update if missing
-	pids, ok := pidsByUser[uid]
-	if !ok {
-		// log.Trace("proc: no processes of user, updating table")
-		updatePids()
-		pidsUpdated = true
-		pids, ok = pidsByUser[uid]
-	}
-	if ok {
-		// if user has pids, start checking them first
-		var checkedUserPids []int
-		for _, possiblePID := range pids {
-			if findSocketFromPid(possiblePID, inode) {
-				return possiblePID
+		// Get all pids for the given uid.
+		pids, ok := getPidsByUser(uid)
+		if !ok {
+			// If we cannot find the uid in the map, update it.
+			updatePids()
+			pidsUpdated = true
+			pids, ok = getPidsByUser(uid)
+		}
+
+		// If we have found PIDs, search them.
+		if ok {
+			for _, pid = range pids {
+				if findSocketFromPid(pid, socketName) {
+					return pid
+				}
 			}
-			checkedUserPids = append(checkedUserPids, possiblePID)
 		}
-		// if we fail on the first run and have not updated, update and check the ones we haven't tried so far.
+
+		// If we still cannot find our socket, and haven't yet updated the PID map,
+		// do this and then check again immediately.
 		if !pidsUpdated {
-			// log.Trace("proc: socket not found in any process of user, updating table")
-			// update
 			updatePids()
-			// sort for faster search
-			for i, j := 0, len(checkedUserPids)-1; i < j; i, j = i+1, j-1 {
-				checkedUserPids[i], checkedUserPids[j] = checkedUserPids[j], checkedUserPids[i]
-			}
-			len := len(checkedUserPids)
-			// check unchecked pids
-			for _, possiblePID := range pids {
-				// only check if not already checked
-				if sort.SearchInts(checkedUserPids, possiblePID) == len {
-					if findSocketFromPid(possiblePID, inode) {
-						return possiblePID
+			pids, ok = getPidsByUser(uid)
+			if ok {
+				for _, pid = range pids {
+					if findSocketFromPid(pid, socketName) {
+						return pid
```

Comment:
This is the exact same code block as above and we would need to keep them in sync. What about doing an `continue`? That would be much more clear to me.